### PR TITLE
Fix converting template types to handle nested templates

### PR DESCRIPTION
**Pull request by existentialism**
Fixes https://github.com/babel/babel-eslint/issues/603 (and the fixture from https://github.com/babel/babel-eslint/issues/609 works).

Reworks our code that converts the format of Babylon template tokens to be a bit more robust, especially with things like nested templates with arrows.

(Adapted the logic from https://github.com/eslint/espree/blob/master/lib/token-translator.js)

### CODE REVIEW

File: lib/babylon-to-espree/index.js [hunk 1]
```diff
@@ -6,11 +6,6 @@ var toTokens = require("./toTokens");
 var toAST = require("./toAST");
 
 module.exports = function(ast, traverse, tt, code) {
-  // remove EOF token, eslint doesn't use this for anything and it interferes
-  // with some rules see https://github.com/babel/babel-eslint/issues/2
-  // todo: find a more elegant way to do this
-  ast.tokens.pop();
```
[the remark this answers is not in the document]
Comment (kaicataldo) [reply lib/babylon-to-espree/index.js]: Hmm, yeah, I'm not sure. What do you mean by
> the new code relies on the EOF token

?

[the remark this answers is not in the document]
Comment (existentialism) [reply lib/babylon-to-espree/index.js]: https://github.com/babel/babel-eslint/pull/610/files#diff-b3374ace9208f796e050e9b907aecf22R74

[the remark this answers is not in the document]
Comment (kaicataldo) [reply lib/babylon-to-espree/index.js]: Ah, sorry, missed that. Thanks!

[the remark this answers is not in the document]
Comment (kaicataldo) [reply lib/babylon-to-espree/index.js]: Could it be removed after the token conversion happens, just to be safe?

[the remark this answers is not in the document]
Comment (existentialism) [reply lib/babylon-to-espree/index.js]: @kaicataldo yep, just updated the PR!

[the remark this answers is not in the document]
Comment (kaicataldo) [reply lib/babylon-to-espree/index.js]: Thinking about this some more, I wonder if we should just remove this. Acorn (the underlying parser for the default parse) also creates these tokens, so seems like it should be fine to remove?

**Pull request merged**